Add checking of Python type hints via mypy in the lints Github Action

Type annotations in Python can be added progressively to a codebase using `mypy` or other tools. Doing so reduces the likelihood of introducing certain kinds of bugs, and adding types to existing code can reveal subtle bugs that already exist. 

It can be introduced using very generous settings to reduce the number of initial type tasks that need to be performed. For example:

```
mypy . --ignore-missing-imports --allow-untyped-globals
```

plus installing some type stubs returns a fairly small number of errors:

```
$ mypy . --ignore-missing-imports --allow-untyped-globals

config/settings/settings_dev.py:45: error: Unsupported target for indexed assignment ("object")
config/settings/settings_pytest.py:18: error: Value of type "object" is not indexable

main/models.py:977: error: Syntax error in type annotation
main/models.py:977: note: Suggestion: Use Tuple[T1, ..., Tn] instead of (T1, ..., Tn)
main/models.py:977: error: Name "legal_doc_fulltext" is not defined
main/models.py:977: error: Name "textblock" is not defined
main/models.py:1041: error: "Type[object]" has no attribute "objects"

# Might consider just skipping checking the admin at first
main/admin.py:295: error: "Callable[[CasebookAdmin, Any], Any]" has no attribute "short_description"
main/admin.py:315: error: "Callable[[SectionAdmin, Any], Any]" has no attribute "short_description"
main/admin.py:336: error: "Callable[[ContentNodeAdmin, Any], Any]" has no attribute "short_description"
main/admin.py:359: error: "Callable[[ResourceAdmin, Any], Any]" has no attribute "short_description"
main/admin.py:375: error: "Callable[[AnnotationsAdmin, Any], Any]" has no attribute "admin_order_field"
main/admin.py:379: error: "Callable[[AnnotationsAdmin, Any], Any]" has no attribute "admin_order_field"
main/admin.py:417: error: "Callable[[TextBlockAdmin, Any], Any]" has no attribute "short_description"
main/admin.py:472: error: "Callable[[UserAdmin, Any], Any]" has no attribute "admin_order_field"
main/admin.py:584: error: "Callable[[LegalDocumentAdmin, Any], Any]" has no attribute "short_description"
```

--- a/.github/workflows/lints.yml
+++ b/.github/workflows/lints.yml
@@ -7,7 +7,7 @@ jobs:
     runs-on: ubuntu-20.04
 
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
       - uses: actions/setup-python@v4
         with:
           python-version: '3.9'

--- a/.github/workflows/lints.yml
+++ b/.github/workflows/lints.yml
@@ -8,7 +8,9 @@ jobs:
 
     steps:
       - uses: actions/checkout@v2
-
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.9'
       - name: flake8
         run: |
           pip install `egrep -o 'flake8==\S+' web/requirements.txt`  # install our version of flake8
@@ -19,3 +21,9 @@ jobs:
         run: |
           pip install `egrep -o 'black==\S+' web/requirements.txt`  # install our version of black
           black --check --diff . # Uses pyproject.toml
+
+      - name: mypy
+        run: |
+          cd web
+          pip install -r requirements.txt  # install a full environment for mypy
+          PYTHONPATH=. mypy

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -9,7 +9,7 @@ jobs:
     if: github.event_name == 'pull_request' || github.repository == 'harvard-lil/h2o'
 
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
 
       ### build docker images locally ###
 

